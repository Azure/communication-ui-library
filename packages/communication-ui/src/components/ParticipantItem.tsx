// Â© Microsoft Corporation. All rights reserved.

import { memberItemContainerStyle, memberItemIsYouStyle, iconsContainerStyle } from './styles/ParticipantItem.styles';
import {
  ContextualMenu,
  DirectionalHint,
  IContextualMenuItem,
  Persona,
  PersonaSize,
  PersonaPresence,
  Stack
} from '@fluentui/react';
import React, { useRef, useState } from 'react';
import { WithErrorHandling } from '../utils/WithErrorHandling';
import { ErrorHandlingProps } from '../providers/ErrorProvider';
import { useTheme } from '@fluentui/react-theme-provider';

/**
 * Props for ParticipantItem component
 */
export interface ParticipantItemProps {
  /** Name of participant */
  name: string;
  /** Optional indicator to show participant is the user */
  isYou?: boolean;
  /** Optional callback returning a JSX element to override avatar */
  onRenderAvatar?: (props?: ParticipantItemProps) => JSX.Element | null;
  /** Optional array of IContextualMenuItem for contextual menu */
  menuItems?: IContextualMenuItem[];
  /** Optional callback returning a JSX element rendered on the right portion of the ParticipantItem. Intended for adding icons. */
  onRenderIcon?: (props?: ParticipantItemProps) => JSX.Element | null;
  /** Optional PersonaPresence to show participant presence. This will not have an effect if property avatar is assigned */
  presence?: PersonaPresence;
}

const ParticipantItemBase = (props: ParticipantItemProps & ErrorHandlingProps): JSX.Element => {
  const { name, isYou, onRenderAvatar, menuItems, onRenderIcon, presence } = props;
  const [clickEvent, setClickEvent] = useState<MouseEvent | undefined>();
  const [menuHidden, setMenuHidden] = useState<boolean>(true);
  const containerRef = useRef<HTMLDivElement>(null);
  const theme = useTheme();

  const showMenu = (clickEvent: React.MouseEvent<HTMLDivElement, MouseEvent>): void => {
    setClickEvent(clickEvent.nativeEvent);
    setMenuHidden(false);
  };

  const hideMenu = (): void => {
    setClickEvent(undefined);
    setMenuHidden(true);
  };

  const avatarToUse = (
    <Persona
      text={name}
      size={PersonaSize.size32}
      presence={presence}
      onRenderPersonaCoin={onRenderAvatar ? () => onRenderAvatar(props) : undefined}
    />
  );
  return (
    <div ref={containerRef} className={memberItemContainerStyle(theme)} onClick={showMenu}>
      {avatarToUse}
      {isYou && <span className={memberItemIsYouStyle}>(you)</span>}
      {onRenderIcon && <Stack className={iconsContainerStyle}>{onRenderIcon(props)}</Stack>}
      {menuItems && (
        <ContextualMenu
          items={menuItems}
          hidden={menuHidden}
          target={clickEvent ?? containerRef}
          onItemClick={hideMenu}
          onDismiss={hideMenu}
          directionalHint={DirectionalHint.bottomLeftEdge}
        />
      )}
    </div>
  );
};

/**
 * Participant Item component representing a participant in Calling or Chat
 * @param props - ParticipantItemProps & ErrorHandlingProps
 */
export const ParticipantItem = (props: ParticipantItemProps & ErrorHandlingProps): JSX.Element =>
  WithErrorHandling(ParticipantItemBase, props);
