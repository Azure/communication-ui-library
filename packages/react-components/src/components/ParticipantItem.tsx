// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

import {
  ContextualMenu,
  DirectionalHint,
  Icon,
  IContextualMenuItem,
  IStyle,
  mergeStyles,
  Persona,
  PersonaPresence,
  PersonaSize,
  Stack,
  Text
} from '@fluentui/react';
import React, { useMemo, useRef, useState } from 'react';
import { useIdentifiers } from '../identifiers';
import { useLocale } from '../localization';
import { useTheme } from '../theming';
import { BaseCustomStyles, OnRenderAvatarCallback } from '../types';
import {
  displayNoneStyle,
  iconContainerStyle,
  iconStyles,
  meContainerStyle,
  menuButtonContainerStyle,
  participantItemContainerStyle,
  participantStateStringStyles
} from './styles/ParticipantItem.styles';
import { _preventDismissOnEvent as preventDismissOnEvent } from '@internal/acs-ui-common';
/* @conditional-compile-remove(one-to-n-calling) */
/* @conditional-compile-remove(PSTN-calls) */
import { ParticipantState } from '../types';
import { useId } from '@fluentui/react-hooks';

/**
 * Fluent styles for {@link ParticipantItem}.
 *
 * @public
 */
export interface ParticipantItemStyles extends BaseCustomStyles {
  /** Styles for the avatar. */
  avatar?: IStyle;
  /** Styles for the (You) string. */
  me?: IStyle;
  /** Styles for the container of the icon. */
  iconContainer?: IStyle;
  /** Styles for the menu. */
  menu?: IStyle;
}

/**
 * Strings of {@link ParticipantItem} that can be overridden.
 *
 * @public
 */
export interface ParticipantItemStrings {
  /** String shown when participant is me */
  isMeText: string;
  /** String shown when hovering over menu button */
  menuTitle: string;
  /** Label for the remove button in participant menu  */
  removeButtonLabel: string;
  /** Label for the sharing icon in participant state stack  */
  sharingIconLabel: string;
  /** Label for the muted icon in participant state stack  */
  mutedIconLabel: string;
  /** placeholder text for participants who does not have a display name*/
  displayNamePlaceholder?: string;
  /* @conditional-compile-remove(one-to-n-calling) */
  /* @conditional-compile-remove(PSTN-calls) */
  /** String shown when `participantState` is `Ringing` */
  participantStateRinging?: string;
  /* @conditional-compile-remove(one-to-n-calling) */
  /* @conditional-compile-remove(PSTN-calls) */
  /** String shown when `participantState` is `Hold` */
  participantStateHold?: string;
  /* @conditional-compile-remove(hide-attendee-name) */
  /** String for the attendee role */
  attendeeRole: string;
}

/**
 * Props for {@link ParticipantItem}.
 *
 * @public
 */
export interface ParticipantItemProps {
  /** Unique User ID of the participant. This `userId` is available in the `onRenderAvatar` callback function */
  userId?: string;
  /** Name of participant. */
  displayName?: string;
  /** Optional indicator to show participant is the user. */
  me?: boolean;
  /** Optional callback returning a JSX element to override avatar. */
  onRenderAvatar?: OnRenderAvatarCallback;
  /** Optional array of IContextualMenuItem for contextual menu. */
  menuItems?: IContextualMenuItem[];
  /** Optional callback returning a JSX element rendered on the right portion of the ParticipantItem. Intended for adding icons. */
  onRenderIcon?: (props?: ParticipantItemProps) => JSX.Element | null;
  /** Optional PersonaPresence to show participant presence. This will not have an effect if property avatar is assigned. */
  presence?: PersonaPresence;
  /**
   * Allows users to pass in an object contains custom CSS styles.
   * @Example
   * ```
   * <ParticipantItem styles={{ root: { background: 'blue' } }} />
   * ```
   */
  styles?: ParticipantItemStyles;
  /**
   * Optional strings to override in component
   */
  strings?: Partial<ParticipantItemStrings>;
  /**
   * Optional callback when component is clicked
   */
  onClick?: (props?: ParticipantItemProps) => void;
  /** Optional value to determine if the tooltip should be shown for participants or not */
  showParticipantOverflowTooltip?: boolean;
  /* @conditional-compile-remove(one-to-n-calling) */
  /* @conditional-compile-remove(PSTN-calls) */
  /**
   * Optional value to determine and display a participants connection status.
   * For example, `Connecting`, `Ringing` etc.
   * The actual text that is displayed is determined by the localized string
   * corresponding to the provided participant state.
   * For example, `strings.participantStateConnecting` will be used if `participantState` is `Connecting`.
   */
  participantState?: ParticipantState;
  /**
   * Optional aria property that prefixes the ParticipantItems aria content
   * Takes in a unique id value of the element you would like to be read before the ParticipantItem.
   */
  ariaLabelledBy?: string;
}

/**
 * Component to render a calling or chat participant.
 *
 * Displays the participant's avatar, displayName and status as well as optional icons and context menu.
 *
 * @public
 */
export const ParticipantItem = (props: ParticipantItemProps): JSX.Element => {
  const {
    userId,
    displayName,
    onRenderAvatar,
    menuItems,
    onRenderIcon,
    presence,
    styles,
    me,
    onClick,
    showParticipantOverflowTooltip
  } = props;
  const [itemHovered, setItemHovered] = useState<boolean>(false);
  const [menuHidden, setMenuHidden] = useState<boolean>(true);
  const containerRef = useRef<HTMLDivElement>(null);
  const theme = useTheme();
  const localeStrings = useLocale().strings.participantItem;
  const ids = useIdentifiers();
  const uniqueId = useId();

  const strings = { ...localeStrings, ...props.strings };
  const participantStateString = participantStateStringTrampoline(props, strings);
  const showMenuIcon = !participantStateString && (itemHovered || !menuHidden) && menuItems && menuItems?.length > 0;

  // For 'me' show empty name so avatar will get 'Person' icon, when there is no name
  const meAvatarText = displayName?.trim() || '';

  const avatarOptions = {
    text: me ? meAvatarText : displayName?.trim() || strings.displayNamePlaceholder,
    size: PersonaSize.size32,
    presence: presence,
    initialsTextColor: 'white',
    showOverflowTooltip: showParticipantOverflowTooltip,
    showUnknownPersonaCoin: !me && (!displayName?.trim() || displayName === strings.displayNamePlaceholder)
  };

  const avatar = onRenderAvatar ? (
    onRenderAvatar(userId ?? '', avatarOptions)
  ) : (
    <Persona
      className={mergeStyles(
        {
          // Prevents persona text from being vertically truncated if a global line height is less than 1.15.
          lineHeight: '1.15rem'
        },
        styles?.avatar
      )}
      {...avatarOptions}
    />
  );

  const meTextStyle = useMemo(
    () => mergeStyles(meContainerStyle, { color: theme.palette.neutralSecondary }, styles?.me),
    [theme.palette.neutralSecondary, styles?.me]
  );
  const contextualMenuStyle = useMemo(
    () => mergeStyles({ background: theme.palette.neutralLighterAlt }, styles?.menu),
    [theme.palette.neutralLighterAlt, styles?.menu]
  );
  const infoContainerStyle = useMemo(
    () =>
      mergeStyles(
        iconContainerStyle,
        { color: theme.palette.neutralSecondary, marginLeft: 'auto' },
        styles?.iconContainer
      ),
    [theme.palette.neutralSecondary, styles?.iconContainer]
  );

  const onDismissMenu = (): void => {
    setItemHovered(false);
    setMenuHidden(true);
  };

  const menuButton = useMemo(
    () => (
      <Stack
        horizontal={true}
        horizontalAlign="end"
        className={mergeStyles(menuButtonContainerStyle, { color: theme.palette.neutralPrimary })}
        title={strings.menuTitle}
        data-ui-id={ids.participantItemMenuButton}
      >
        <Icon
          iconName="ParticipantItemOptionsHovered"
          className={mergeStyles(iconStyles, !showMenuIcon ? displayNoneStyle : {})}
        />
      </Stack>
    ),
    [theme.palette.neutralPrimary, strings.menuTitle, ids.participantItemMenuButton, showMenuIcon]
  );

  return (
    <div
      ref={containerRef}
      role={'menuitem'}
      aria-disabled={(menuItems && menuItems.length > 0) || props.onClick ? false : true}
      data-is-focusable={true}
      data-ui-id="participant-item"
      className={mergeStyles(
        participantItemContainerStyle({ clickable: !!menuItems && menuItems.length > 0 }, theme),
        styles?.root
      )}
      onMouseEnter={() => setItemHovered(true)}
      onMouseLeave={() => setItemHovered(false)}
      onClick={() => {
        if (!participantStateString) {
          setItemHovered(true);
          setMenuHidden(false);
          onClick?.(props);
        }
        if (!menuHidden) {
          onDismissMenu();
        }
      }}
      tabIndex={0}
    >
      <Stack
        horizontal
        className={mergeStyles({
          flexGrow: 1,
          maxWidth: '100%',
          alignItems: 'center'
        })}
        id={uniqueId}
        aria-labelledby={`${props.ariaLabelledBy} ${uniqueId}`}
      >
        {avatar}
        {me && <Text className={meTextStyle}>{strings.isMeText}</Text>}
        <Stack horizontal className={mergeStyles(infoContainerStyle)}>
          {!showMenuIcon && onRenderIcon && onRenderIcon(props)}
          {/* When the participantStateString has a value, we don't show the menu  */}
          {!me && participantStateString ? (
            <Text data-ui-id="participant-item-state-string" className={mergeStyles(participantStateStringStyles)}>
              {participantStateString}
            </Text>
          ) : (
            <>
              {menuItems && menuItems.length > 0 && (
                <>
                  {menuButton}
                  <ContextualMenu
                    items={menuItems}
                    hidden={menuHidden}
                    target={containerRef}
                    onItemClick={onDismissMenu}
                    onDismiss={onDismissMenu}
                    directionalHint={DirectionalHint.bottomRightEdge}
                    className={contextualMenuStyle}
                    calloutProps={{
                      preventDismissOnEvent
                    }}
                  />
                </>
              )}
            </>
          )}
        </Stack>
      </Stack>
    </div>
  );
};

const participantStateStringTrampoline = (
  props: ParticipantItemProps,
  strings: ParticipantItemStrings
): string | undefined => {
  /* @conditional-compile-remove(one-to-n-calling) */
  /* @conditional-compile-remove(PSTN-calls) */
  return props.participantState === 'EarlyMedia' || props.participantState === 'Ringing'
    ? strings?.participantStateRinging
    : props.participantState === 'Hold'
      ? strings?.participantStateHold
      : undefined;

  return undefined;
};
