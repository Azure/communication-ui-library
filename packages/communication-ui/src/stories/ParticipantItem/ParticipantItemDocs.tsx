// Â© Microsoft Corporation. All rights reserved.

import React from 'react';
import { Title, Description, Props, Heading, Source, Canvas } from '@storybook/addon-docs/blocks';
import { ParticipantItem, ParticipantItemProps } from '../../components';
import { IContextualMenuItem, PersonaPresence, Icon } from '@fluentui/react';

const importStatement = `
import { ParticipantItem } from '@azure/communication-ui';
import { IContextualMenuItem, PersonaPresence } from '@fluentui/react';`;

const ParticipantItemExample: () => JSX.Element = () => {
  const menuItems: IContextualMenuItem[] = [
    {
      key: 'Mute',
      text: 'Mute',
      onClick: () => alert('Mute')
    },
    {
      key: 'Remove',
      text: 'Remove',
      onClick: () => alert('Remove')
    }
  ];

  return <ParticipantItem name="Johnny Bravo" menuItems={menuItems} presence={PersonaPresence.online} />;
};

const exampleCode = `
const menuItems: IContextualMenuItem[] = [
  {
    key: 'Mute',
    text: 'Mute',
    onClick: () => alert('Mute')
  },
  {
    key: 'Remove',
    text: 'Remove',
    onClick: () => alert('Remove')
  }
];

return (
  <ParticipantItem
    name="Johnny Bravo"
    menuItems={menuItems}
    presence={PersonaPresence.online}
  />
);
`;

const CustomAvatarExample: () => JSX.Element = () => {
  const onRenderAvatar = (): JSX.Element => {
    return (
      <img
        src="https://static2.sharepointonline.com/files/fabric/office-ui-fabric-react-assets/persona-female.png"
        width="32px"
        height="32px"
        style={{
          borderRadius: 20,
          display: 'block'
        }}
      />
    );
  };
  return (
    <div style={{ width: '200px' }}>
      <ParticipantItem name="Annie Lindqvist" onRenderAvatar={onRenderAvatar} />
    </div>
  );
};

const customAvatarCode = `
const onRenderAvatar = (): JSX.Element => {
  return (
    <img
      src="https://static2.sharepointonline.com/files/fabric/office-ui-fabric-react-assets/persona-female.png"
      width="32px"
      height="32px"
      style={{
        borderRadius: 20,
        display: 'block'
      }}
    />
  );
};
<div style={{ width: '200px' }}>
  <ParticipantItem name="Annie Lindqvist" onRenderAvatar={onRenderAvatar} />
</div>
`;

const CustomIconExample: () => JSX.Element = () => {
  const onRenderIcon = (props?: ParticipantItemProps): JSX.Element | null => {
    // eslint-disable-next-line react/prop-types
    if (props?.name === 'Patrick') {
      return <Icon iconName="FavoriteStar" />;
      // eslint-disable-next-line react/prop-types
    } else if (props?.isYou) {
      return null;
    }
    return <Icon iconName="AddFriend" />;
  };
  return (
    <div style={{ width: '200px' }}>
      <ParticipantItem name="Spongebob" isYou={true} onRenderIcon={onRenderIcon} />
      <ParticipantItem name="Patrick" onRenderIcon={onRenderIcon} />
      <ParticipantItem name="Sandy" onRenderIcon={onRenderIcon} />
    </div>
  );
};

const customIconCode = `
const onRenderIcon = (props?: ParticipantItemProps): JSX.Element | null => {
  if (props?.name === 'Patrick') {
    return <Icon iconName="FavoriteStar" />;
  } else if (props?.isYou) {
    return null;
  }
  return <Icon iconName="AddFriend" />;
};
return (
  <div style={{ width: '200px' }}>
    <ParticipantItem name="Spongebob" isYou={true} onRenderIcon={onRenderIcon} />
    <ParticipantItem name="Patrick" onRenderIcon={onRenderIcon} />
    <ParticipantItem name="Sandy" onRenderIcon={onRenderIcon} />
  </div>
);
`;

export const getDocs: () => JSX.Element = () => {
  return (
    <>
      <Title>ParticipantItem</Title>
      <Description>
        The ParticipantItem component represents a user and displays their avatar, name, status and additional icons.
      </Description>
      <Heading>Importing</Heading>
      <Source code={importStatement} />
      <Heading>Example</Heading>
      <Canvas>
        <ParticipantItemExample />
      </Canvas>
      <Source code={exampleCode} />
      <Heading>Custom avatar</Heading>
      To customize the avatar, use the onRenderAvatar property like in the example below. Note: the avatar element is
      recommended to be within 32 by 32 pixels.
      <Source code={customAvatarCode} />
      <Canvas>
        <CustomAvatarExample />
      </Canvas>
      <Heading>Add icon</Heading>
      To add an icon, use the onRenderIcon property like in the example below.
      <Source code={customIconCode} />
      <Canvas>
        <CustomIconExample />
      </Canvas>
      <Heading>Props</Heading>
      <Props of={ParticipantItem} />
    </>
  );
};
