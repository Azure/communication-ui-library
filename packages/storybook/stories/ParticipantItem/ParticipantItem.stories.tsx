// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.

// also exported from '@storybook/react' if you can deal with breaking changes in 6.1
import { ParticipantItem as ParticipantItemComponent } from '@azure/communication-react';
import { Stack, mergeStyles } from '@fluentui/react';
import { MicOff16Regular, ShareScreenStart20Filled } from '@fluentui/react-icons';
import { Title, Description, Props, Heading, Source, Canvas } from '@storybook/addon-docs';
import { Meta } from '@storybook/react/types-6-0';
import React from 'react';

import { COMPONENT_FOLDER_PREFIX } from '../constants';
import { controlsToAdd, hiddenControl } from '../controlsUtils';
import { CustomAvatarExample } from './snippets/CustomAvatar.snippet';
import { CustomIconExample } from './snippets/CustomIcon.snippet';
import { ParticipantItemExample } from './snippets/ParticipantItem.snippet';

const CustomAvatarExampleText = require('!!raw-loader!./snippets/CustomAvatar.snippet.tsx').default;
const CustomIconExampleText = require('!!raw-loader!./snippets/CustomIcon.snippet.tsx').default;
const ParticipantItemExampleText = require('!!raw-loader!./snippets/ParticipantItem.snippet.tsx').default;

const importStatement = `
import { ParticipantItem, ParticipantItemProps } from '@azure/communication-react';
import { IContextualMenuItem, PersonaPresence } from '@fluentui/react';`;

const getDocs: () => JSX.Element = () => {
  return (
    <>
      <Title>ParticipantItem</Title>
      <Description>
        Component to render a calling or chat participant. This component can display the participant's avatar,
        displayName, online presence as well as optional icons and context menu.
      </Description>

      <Heading>Importing</Heading>
      <Source code={importStatement} />

      <Heading>Usage</Heading>
      <Description>
        Here is an example of how to use `ParticipantItem.` In this example, the `menuItems` property is used to add a
        context menu. The type of `menuItems` is an array of
        [IContextualMenuItem](https://developer.microsoft.com/fluentui#/controls/web/contextualmenu#IContextualMenuItem).
        Click on the rendered participant below to see the menu items.
      </Description>
      <Canvas mdxSource={ParticipantItemExampleText}>
        <ParticipantItemExample />
      </Canvas>

      <Heading>Custom avatar</Heading>
      <Description>
        To customize the avatar, use the `onRenderAvatar` property like in the example below. We recommend the avatar
        element to be within 32 by 32 pixels.
      </Description>
      <Canvas mdxSource={CustomAvatarExampleText}>
        <CustomAvatarExample />
      </Canvas>

      <Heading>Add icon</Heading>
      <Description>To add an icon, use the `onRenderIcon` property like in the example below.</Description>
      <Canvas mdxSource={CustomIconExampleText}>
        <CustomIconExample />
      </Canvas>

      <Heading>Props</Heading>
      <Props of={ParticipantItemComponent} />
    </>
  );
};

const onlyUnique = (value: string, index: number, self: string[]): boolean => {
  return self.indexOf(value) === index;
};

const ParticipantItemStory: (args) => JSX.Element = (args) => {
  const menuItems = args.menuItemsStr
    .split(',')
    .map((menuItem) => menuItem.trim())
    .filter(onlyUnique)
    .map((menuItem) => {
      return {
        key: menuItem,
        text: menuItem
      };
    });

  const containerStyle = { width: '15rem' };
  const iconStyles = mergeStyles({ display: 'flex', alignItems: 'center' });
  const tokenProps = { childrenGap: '0.5rem' };

  return (
    <div style={containerStyle}>
      <ParticipantItemComponent
        displayName={args.displayName}
        me={args.me}
        menuItems={menuItems}
        onRenderIcon={() => (
          <Stack horizontal tokens={tokenProps}>
            {args.isScreenSharing && <ShareScreenStart20Filled className={iconStyles} primaryFill="currentColor" />}
            {args.isMuted && <MicOff16Regular className={iconStyles} primaryFill="currentColor" />}
          </Stack>
        )}
      />
    </div>
  );
};

// This must be the only named export from this module, and must be named to match the storybook path suffix.
// This ensures that storybook hoists the story instead of creating a folder with a single entry.
export const ParticipantItem = ParticipantItemStory.bind({});

export default {
  id: `${COMPONENT_FOLDER_PREFIX}-participantitem`,
  title: `${COMPONENT_FOLDER_PREFIX}/Participant Item`,
  component: ParticipantItemComponent,
  argTypes: {
    displayName: { ...controlsToAdd.displayName, defaultValue: 'August Manning' },
    isScreenSharing: { ...controlsToAdd.isScreenSharing, defaultValue: true },
    isMuted: { ...controlsToAdd.isMuted, defaultValue: true },
    me: controlsToAdd.isMe,
    menuItemsStr: controlsToAdd.participantItemMenuItemsStr,
    // Hiding auto-generated controls
    onRenderAvatar: hiddenControl,
    menuItems: hiddenControl,
    onRenderIcon: hiddenControl,
    presence: hiddenControl,
    styles: hiddenControl,
    strings: hiddenControl,
    userId: hiddenControl
  },
  parameters: {
    docs: {
      page: () => getDocs()
    }
  }
} as Meta;
