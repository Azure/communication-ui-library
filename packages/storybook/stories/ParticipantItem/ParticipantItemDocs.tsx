// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.

import { ParticipantItem } from '@azure/communication-react';
import { Title, Description, Props, Heading, Source, Canvas } from '@storybook/addon-docs/blocks';
import React from 'react';
import { CustomAvatarExample } from './snippets/CustomAvatar.snippet';
import { CustomIconExample } from './snippets/CustomIcon.snippet';
import { ParticipantItemExample } from './snippets/ParticipantItem.snippet';

const CustomAvatarExampleText = require('!!raw-loader!./snippets/CustomAvatar.snippet.tsx').default;
const CustomIconExampleText = require('!!raw-loader!./snippets/CustomIcon.snippet.tsx').default;
const ParticipantItemExampleText = require('!!raw-loader!./snippets/ParticipantItem.snippet.tsx').default;

const importStatement = `
import { ParticipantItem, ParticipantItemProps } from '@azure/communication-react';
import { IContextualMenuItem, PersonaPresence } from '@fluentui/react';`;

export const getDocs: () => JSX.Element = () => {
  return (
    <>
      <Title>ParticipantItem</Title>
      <Description of={ParticipantItem} />
      <Heading>Importing</Heading>
      <Source code={importStatement} />
      <Heading>Usage</Heading>
      <Description>
        Here is an example of how to use `ParticipantItem.` In this example, the `menuItems` property is used to add a
        context menu. The type of `menuItems` is an array of
        [IContextualMenuItem](https://developer.microsoft.com/en-us/fluentui#/controls/web/contextualmenu#IContextualMenuItem).
        Click on the rendered participant below to see the menu items.
      </Description>
      <Canvas withSource={'none' as any}>
        <ParticipantItemExample />
      </Canvas>
      <Source code={ParticipantItemExampleText} />
      <Heading>Custom avatar</Heading>
      <Description>
        To customize the avatar, use the `onRenderAvatar` property like in the example below. We recommend the avatar
        element to be within 32 by 32 pixels.
      </Description>
      <Source code={CustomAvatarExampleText} />
      <Canvas withSource={'none' as any}>
        <CustomAvatarExample />
      </Canvas>
      <Heading>Add icon</Heading>
      <Description>To add an icon, use the `onRenderIcon` property like in the example below.</Description>
      <Source code={CustomIconExampleText} />
      <Canvas withSource={'none' as any}>
        <CustomIconExample />
      </Canvas>
      <Heading>Props</Heading>
      <Props of={ParticipantItem} />
    </>
  );
};
