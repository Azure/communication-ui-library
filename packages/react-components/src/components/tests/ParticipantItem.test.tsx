// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.

import React from 'react';
import createComponentWithLocalization from './createComponentWithLocalization';
import { ParticipantItem } from '../ParticipantItem';
import englishStrings from '../../localization/translated/en-US.json';

test('ParticipantItem snapshot in en-US', async () => {
  const component = createComponentWithLocalization(<ParticipantItem displayName="Mark" me={true} />, {
    locale: { locale: 'en-US', rtl: false, strings: englishStrings }
  });

  expect(component).toMatchSnapshot();
});
