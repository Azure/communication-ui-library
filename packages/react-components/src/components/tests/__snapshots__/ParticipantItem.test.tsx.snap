// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ParticipantItem snapshot in en-US 1`] = `
<body
  className="ms-Fabric"
  dir="ltr"
>
  <div
    className="css-53"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <div
      className="ms-Persona ms-Persona--size32 root-60"
    >
      <div
        className="ms-Persona-coin ms-Persona--size32 coin-67"
        role="presentation"
      >
        <div
          className="ms-Persona-imageArea imageArea-69"
          role="presentation"
        >
          <div
            aria-hidden="true"
            className="ms-Persona-initials initials-72"
          >
            <span>
              M
            </span>
          </div>
        </div>
      </div>
      <div
        className="ms-Persona-details details-61"
      >
        <div
          className="ms-Persona-primaryText primaryText-62"
          dir="auto"
        >
          <div
            className="ms-TooltipHost root-75"
            onBlurCapture={[Function]}
            onFocusCapture={[Function]}
            onKeyDown={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
          >
            Mark
          </div>
        </div>
        <div
          className="ms-Persona-secondaryText secondaryText-63"
          dir="auto"
        />
        <div
          className="ms-Persona-tertiaryText tertiaryText-64"
          dir="auto"
        />
        <div
          className="ms-Persona-optionalText optionalText-65"
          dir="auto"
        />
      </div>
    </div>
    <div
      className="ms-Stack css-76"
    />
  </div>
</body>
`;
